### PR TITLE
feat: BBS+ signature verification

### PR DESCRIPTION
**Pull request by kdimak**
Signed-off-by: Dmitriy Kinoshenko <dkinoshenko@gmail.com>

#2221

Follow-up PRs:
- JSON-LD BbsBlsSignature2020 signature suite
- support of VC/VP verification

### CODE REVIEW

File: pkg/crypto/bls/bbs12381g2pub/public_key.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+func ParsePublicKey(pubKeyBytes []byte) (*PublicKey, error) {
+	if len(pubKeyBytes) != bls12381G2PublicKeyLen {
+		return nil, errors.New("invalid size of public key")
+	}
+
+	var pkBytesArr [bls12381G2PublicKeyLen]byte
+
+	copy(pkBytesArr[:], pubKeyBytes[:bls12381G2PublicKeyLen])
+
+	publicKey, err := g2pubs.DeserializePublicKey(pkBytesArr)
```
Comment (mikelodder7) [inline pkg/crypto/bls/bbs12381g2pub/public_key.go:32]: Does `DeserializePublicKey` check to make sure the public key is not infinity and is on the curve?

File: pkg/crypto/bls/bbs12381g2pub/signature.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	var sigBytesArr [g1CompressedSize]byte
+
+	copy(sigBytesArr[:], sigBytes[:g1CompressedSize])
+
+	signature, err := g2pubs.DeserializeSignature(sigBytesArr)
+	if err != nil {
+		return nil, fmt.Errorf("deserialize G1 compressed signature: %w", err)
+	}
+
+	e := parseFr(sigBytes[g1CompressedSize : g1CompressedSize+frCompressedSize])
```
Comment (mikelodder7) [inline pkg/crypto/bls/bbs12381g2pub/signature.go:40]: check that `e` and `s` are not zero

File: pkg/crypto/bls/bbs12381g2pub/signature.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+func ParseSignature(sigBytes []byte) (*Signature, error) {
+	if len(sigBytes) != bls12381SignatureLen {
+		return nil, errors.New("invalid size of signature")
+	}
+
+	var sigBytesArr [g1CompressedSize]byte
+
+	copy(sigBytesArr[:], sigBytes[:g1CompressedSize])
+
+	signature, err := g2pubs.DeserializeSignature(sigBytesArr)
```
Comment (mikelodder7) [inline pkg/crypto/bls/bbs12381g2pub/signature.go:35]: Does `DeserializeSignature` check if the point is not at infinity?

File: pkg/crypto/bls/bbs12381g2pub/bls.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+		dataCopy := make([]byte, len(data))
+		copy(dataCopy, data)
+
+		iBytes := uint32ToBytes(uint32(i))
+
+		for j := 0; j < len(iBytes); j++ {
+			dataCopy[j+offset] = iBytes[j]
+		}
+
+		h[i-1], err = hashToG1(dataCopy)
```
[the remark this answers is not in the document]
Comment (kdimak) [reply pkg/crypto/bls/bbs12381g2pub/bls.go:115]: I re-checked it, it's OK. We set the values of `h` slice at line 115, and then assign them to `bases` at line 125.

File: pkg/crypto/bls/bbs12381g2pub/public_key.go (hunk 1 above)
Comment (kdimak) [reply pkg/crypto/bls/bbs12381g2pub/public_key.go:32]: Yes, it checks.

File: pkg/crypto/bls/bbs12381g2pub/signature.go (hunk 3 above)
Comment (kdimak) [reply pkg/crypto/bls/bbs12381g2pub/signature.go:35]: Yes, it checks (the same as for `DeserializePublicKey()`).

File: pkg/crypto/bls/bbs12381g2pub/signature.go (hunk 2 above)
Comment (kdimak) [reply pkg/crypto/bls/bbs12381g2pub/signature.go]: Added a check at `parseFr()` that now returns an error at a second value.

File: pkg/crypto/bls/bbs12381g2pub/bls.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+		if err != nil {
+			return nil, fmt.Errorf("create G1 point from hash")
```
Comment (Baha-sk) [inline pkg/crypto/bls/bbs12381g2pub/bls.go:120]: should `err` be returned here?

**Review by fqutishat — APPROVED** [review]

File: go.mod [hunk 6]
```diff
@@ -55,6 +57,8 @@ require (
 replace (
 	github.com/piprate/json-gold => github.com/trustbloc/json-gold v0.3.1-0.20200414173446-30d742ee949e
 	golang.org/x/sys => golang.org/x/sys v0.0.0-20200826173525-f9321e4c35a6
+	github.com/kilic/bls12-381 => github.com/trustbloc/bls12-381 v0.0.0-20201008080608-ba2e87ef05ef
```
Comment (troyronda) [inline go.mod:60]: We need an issue to remove these forked references.

File: go.mod [hunk 7]
```diff
@@ -55,6 +57,8 @@ require (
...
+	github.com/phoreproject/bls => github.com/trustbloc/bls v0.0.0-20201008085849-81064514c3cc
```
Comment (troyronda) [inline go.mod:61]: We need an issue to remove this dependency.

File: pkg/crypto/bls/api.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package bls
+
+// Bls defines BBS+ signature scheme (https://eprint.iacr.org/2016/663.pdf, section 4.3).
+type Bls interface {
```
Comment (troyronda) [inline pkg/crypto/bls/api.go:10]: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: pkg/crypto/bls/bbs12381g2pub/bls.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+	"errors"
+	"fmt"
+	"hash"
+
+	bls12381 "github.com/kilic/bls12-381"
+	"github.com/phoreproject/bls"
+	"golang.org/x/crypto/blake2b"
+)
+
+// BlsG2Pub defines BBS+ signature scheme where public key is a point in the field of G2.
```
Comment (troyronda) [inline pkg/crypto/bls/bbs12381g2pub/bls.go:20]: https://github.com/golang/go/wiki/CodeReviewComments#initialisms

File: pkg/crypto/bls/bbs12381g2pub/bls.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+type BlsG2Pub struct {
+}
+
+// NewBlsG2Pub creates a new BlsG2Pub.
+func NewBlsG2Pub() *BlsG2Pub {
```
Comment (troyronda) [inline pkg/crypto/bls/bbs12381g2pub/bls.go:25]: Do you plan other constructors in this package?

File: pkg/crypto/bls/bbs12381g2pub/bls.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	binary.BigEndian.PutUint32(bytes, value)
+
+	return bytes
+}
+
+func hashToG1(data []byte) (*bls.G1Projective, error) {
+	dstG1 := []byte("BLS12381G1_XMD:BLAKE2B_SSWU_RO_BBS+_SIGNATURES:1_0_0")
+
+	newBlake2b := func() hash.Hash {
+		h, _ := blake2b.New512(nil) //nolint:errcheck
```
Comment (troyronda) [inline pkg/crypto/bls/bbs12381g2pub/bls.go:172]: please add comment on why error doesn't need to be checked.

File: pkg/crypto/bls/bbs12381g2pub/bls.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+	p0Bls.SetRawBytes(p0BytesArr)
+
+	return p0Bls.ToProjective(), nil
+}
+
+func compareTwoPairings(p1 *bls.G1Projective, q1 *bls.G2Projective, p2 *bls.G1Projective, q2 *bls.G2Projective) bool {
+	engine := bls12381.NewEngine()
+
+	//nolint:errcheck
```
Comment (troyronda) [inline pkg/crypto/bls/bbs12381g2pub/bls.go:199]: please add comment on why error doesn't need to be checked.

File: go.mod (hunk 6 above)
Comment (kdimak) [reply go.mod:60]: https://github.com/hyperledger/aries-framework-go/issues/2247

File: go.mod (hunk 7 above)
Comment (kdimak) [reply go.mod:61]: https://github.com/hyperledger/aries-framework-go/issues/2248

File: pkg/crypto/bls/api.go (hunk 8 above)
Comment (kdimak) [reply pkg/crypto/bls/api.go:10]: Renamed to `BBS`

File: pkg/crypto/bls/bbs12381g2pub/bls.go (hunk 9 above)
Comment (kdimak) [reply pkg/crypto/bls/bbs12381g2pub/bls.go:20]: Renamed to `BBSG2Pub`.

File: pkg/crypto/bls/bbs12381g2pub/bls.go (hunk 10 above)
Comment (kdimak) [reply pkg/crypto/bls/bbs12381g2pub/bls.go:25]: Actually no. There are several `Parse...()` methods only. So yes - renaming to just `New()`.

**Review by Baha-sk — APPROVED** [review]

**Comment by kdimak** [comment]
@Baha-sk @troyronda The package is moved from `pkg/crypto` to `pkg/doc`. Also, it's renamed from `bls` to `bbs` (to fit the BBS+ signature schema better).

File: pkg/doc/bbs/bbs12381g2pub/bbs.go [hunk 13]
```diff
@@ -0,0 +1,230 @@
+/*
```
Comment (troyronda) [inline pkg/doc/bbs/bbs12381g2pub/bbs.go:1]: @kdimak I'm not understanding why BBS is at the top-level of pkg/doc .

**Review by troyronda — APPROVED** [review]
We should discuss the location of these packages further.

**Pull request merged**